subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/AbstractSourceIncrementalCompilationIntegrationTest.groovy

```diff
@@ -398,6 +398,33 @@ abstract class AbstractSourceIncrementalCompilationIntegrationTest extends Abstr
         outputs.recompiledClasses("B", "A")
     }
 
+    def 'can move classes between source dirs'() {
+        given:
+        buildFile << "sourceSets.main.${language.name}.srcDir 'extra'"
+        source('class A1 {}')
+        file("extra/A2.${language.name}") << "class A2 {}"
+        file("extra/B.${language.name}") << "class B {}"
```

Comment:
Can we add support for declaring packages? I think this may behave differently when not in the root directory.

Moreover, let's declare the classes public, so the file name needs to be the same as the class name, and change inner classes.

Reply:
Here is my suggestion how to make the test better: https://github.com/gradle/gradle/commit/f324a8f67ccd022a5a7dade06175a76cb9b6fa2c

Reply:
Thanks. Already copied your change. Actually `TestFile` has a `moveToDir` method so we don't need to delete/recreate the file.